### PR TITLE
[devtools-engineering] Debugging WebAssembly Faster

### PR DESCRIPTION
**Pull request by jecfish**

### CODE REVIEW

**Comment by sbc100** [comment]
Here are my co-author details:
```
$ git diff
diff --git a/site/_data/authorsData.json b/site/_data/authorsData.json
index ad577fb5..34900433 100644
--- a/site/_data/authorsData.json
+++ b/site/_data/authorsData.json
@@ -697,5 +697,9 @@
     "github": "twerske",
     "twitter": "twerske",
     "image": "image/jxu1OdD7LKOGIDU7jURMpSH2lyK2/7lFlJupvNAPBh7huFYyg.png"
+  },
+  "sbc": {
+    "country": "US",
+    "github": "sbc100"
   }
 }
diff --git a/site/_data/i18n/authors.yaml b/site/_data/i18n/authors.yaml
index d6768e53..9032a30a 100644
--- a/site/_data/i18n/authors.yaml
+++ b/site/_data/i18n/authors.yaml
@@ -785,3 +785,8 @@ emmatwersky:
     en: 'Emma Twersky'
   description:
     en: 'Developer Relations Engineer, Google'
+sbc:
+  title:
+    en: 'Sam Clegg'
+  description:
+    en: 'Software engineer, Google'
```

Let me know if you want me to send that as a separate PR.

File: site/_data/authorsData.json [hunk 1]
```diff
@@ -697,5 +697,10 @@
     "github": "twerske",
     "twitter": "twerske",
     "image": "image/jxu1OdD7LKOGIDU7jURMpSH2lyK2/7lFlJupvNAPBh7huFYyg.png"
+  },
+  "pfaffe": {
+    "country": "DE",
+    "github": "pfaffe",
+    "image": "image/dPDCek3EhZgLQPGtEG3y0fTn4v82/sTBqtIKOD5huM7ojQ375.jpg"
```
Comment (jecfish) [inline site/_data/authorsData.json:704]:
```suggestion
    "image": ""
```

File: site/en/blog/faster-wasm-debugging/index.md [hunk 2]
```diff
@@ -0,0 +1,227 @@
+---
+title: "Debugging WebAssembly Faster"
+description: >
+  How we make debugging WebAssembly faster for large applications.
+layout: "layouts/blog-post.njk"
+authors:
+  - pfaffe
```
Comment (jecfish) [inline site/en/blog/faster-wasm-debugging/index.md:7]:
```suggestion
  - pfaffe
  - kimanh
```

File: site/_data/i18n/authors.yaml [hunk 3]
```diff
@@ -785,3 +785,8 @@ emmatwersky:
     en: 'Emma Twersky'
   description:
     en: 'Developer Relations Engineer, Google'
+pfaffe:
+  title:
+    en: 'Philip Pfaffe'
+  description:
+    en: 'Software Engineer working on Chrome DevTools at Google.'
```
Comment (jecfish) [inline site/_data/i18n/authors.yaml:792]:
```suggestion
    en: 'Software Engineer working on Chrome DevTools at Google.'
sbc:
  title:
    en: 'Sam Clegg'
    description:
      en: 'Software engineer, Google'
```

File: site/_data/authorsData.json [hunk 4]
```diff
@@ -697,5 +697,10 @@
     "github": "twerske",
     "twitter": "twerske",
     "image": "image/jxu1OdD7LKOGIDU7jURMpSH2lyK2/7lFlJupvNAPBh7huFYyg.png"
+  },
+  "pfaffe": {
+    "country": "DE",
+    "github": "pfaffe",
+    "image": ""
   }
```
Comment (jecfish) [inline site/_data/authorsData.json:705]:
```suggestion
  },
  "sbc": {
    "country": "US",
    "github": "sbc100"
   }
```

**Comment by jecfish** [comment]
Thanks @sbc100! I 've added that in this PR

File: site/_data/authorsData.json [hunk 5]
```diff
@@ -697,5 +697,14 @@
...
+  },
+  "pfaffe": {
+    "country": "DE",
+    "github": "pfaffe",
+    "image": ""
+  },
+  "sbc": {
+    "country": "US",
+    "github": "sbc100"
+   }
```
Comment (jecfish) [inline site/_data/authorsData.json:709]:
```suggestion
   },
   "ericsl": {
     "country": "DE",
     "github": "ericsl"
   }
```

File: site/_data/i18n/authors.yaml [hunk 6]
```diff
@@ -785,3 +785,13 @@ emmatwersky:
...
+sbc:
+  title:
+    en: 'Sam Clegg'
+    description:
+      en: 'Software engineer, Google'
```
Comment (jecfish) [inline site/_data/i18n/authors.yaml:797]:
```suggestion
      en: 'Software engineer, Google'
ericsl:
  title:
    en: 'Eric Leese'
  description:
    en: 'Software Engineer working on Chrome DevTools at Google.'
```

File: site/en/blog/faster-wasm-debugging/index.md [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+  - kimanh
```
Comment (jecfish) [inline site/en/blog/faster-wasm-debugging/index.md:8]:
```suggestion
  - kimanh
  - ericsl
  - sbc
```

File: site/_data/authorsData.json [hunk 8]
```diff
@@ -697,5 +697,18 @@
     "github": "twerske",
     "twitter": "twerske",
     "image": "image/jxu1OdD7LKOGIDU7jURMpSH2lyK2/7lFlJupvNAPBh7huFYyg.png"
-  }
+  },
+  "pfaffe": {
+    "country": "DE",
+    "github": "pfaffe",
+    "image": ""
```
Comment (jecfish) [inline site/_data/authorsData.json:704]:
```suggestion
```

**Comment by jecfish** [comment]
@sofiayem please help to ruuberstamp this

File: site/_data/authorsData.json [hunk 9]
```diff
@@ -697,5 +697,17 @@
     "github": "twerske",
     "twitter": "twerske",
     "image": "image/jxu1OdD7LKOGIDU7jURMpSH2lyK2/7lFlJupvNAPBh7huFYyg.png"
-  }
+  },
+  "pfaffe": {
+    "country": "DE",
+    "github": "pfaffe",
```
Comment (jecfish) [inline site/_data/authorsData.json:703]:
```suggestion
    "github": "pfaffe"
```

File: site/en/blog/faster-wasm-debugging/index.md [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+  // Render everything we've drawn to the canvas.
+  SDL_RenderPresent(renderer);
+
+  // SDL_Quit();
+}
+```
+
+It's still a fairly small example and you would likely not see any of the real issues you would see in a really big application, but we can still show you what the new features are. It's quick and easy to set up and to try for yourself!
+
+In the last post, we discussed how to compile and debug this example. Let's do that again, but let's also take a peak at the //performance//:
```
Comment (pfaffe) [inline site/en/blog/faster-wasm-debugging/index.md:83]: //performance// should be italic.

File: site/en/blog/faster-wasm-debugging/index.md [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+
+```bash
+$ emcc -sUSE_SDL=2 -g -O0 -o mandelbrot.html mandelbrot.cc -sALLOW_MEMORY_GROWTH
+```
+
+This command produces a 3MB wasm binary. And the bulk of that, as you might expect, is debug information. You can verify this with the `llvm-objdump` tool[^1] for example:
+
+[^1]: If you don't have a recent version of `llvm-objdump` on your system, and you are using `emsdk`, you can find it in the `emsdk/upstream/bin` directory.
```
Comment (pfaffe) [inline site/en/blog/faster-wasm-debugging/index.md:91]: Apparently the footnote didn't work, is there a way to do that?

File: site/en/blog/faster-wasm-debugging/index.md (hunk 10 above)
Comment (jecfish) [inline site/en/blog/faster-wasm-debugging/index.md:83]:
```suggestion
In the last post, we discussed how to compile and debug this example. Let's do that again, but let's also take a peak at the _//performance//_:
```

File: site/en/blog/faster-wasm-debugging/index.md (hunk 11 above)
Comment (jecfish) [inline site/en/blog/faster-wasm-debugging/index.md:91]:
```suggestion
[<span id="footnote-1">1</span>]: If you don't have a recent version of `llvm-objdump` on your system, and you are using `emsdk`, you can find it in the `emsdk/upstream/bin` directory.
```

File: site/en/blog/faster-wasm-debugging/index.md [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+
+It's still a fairly small example and you would likely not see any of the real issues you would see in a really big application, but we can still show you what the new features are. It's quick and easy to set up and to try for yourself!
+
+In the last post, we discussed how to compile and debug this example. Let's do that again, but let's also take a peak at the //performance//:
+
+```bash
+$ emcc -sUSE_SDL=2 -g -O0 -o mandelbrot.html mandelbrot.cc -sALLOW_MEMORY_GROWTH
+```
+
+This command produces a 3MB wasm binary. And the bulk of that, as you might expect, is debug information. You can verify this with the `llvm-objdump` tool[^1] for example:
```
Comment (jecfish) [inline site/en/blog/faster-wasm-debugging/index.md:89]:
```suggestion
This command produces a 3MB wasm binary. And the bulk of that, as you might expect, is debug information. You can verify this with the `llvm-objdump` tool< sup >[[1]](#footnote-1)</sup> for example:
```

File: site/en/blog/faster-wasm-debugging/index.md [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+
+Building the DWARF package out of the individual objects has the advantage that you only need to serve one extra file! We're currently working on also loading all the individual objects in a future release.
+
+
+### What's with DWARF 5?
+
+You might have noticed, we snuck another flag into the `emcc` command above, `-gdwarf-5`. Enabling version 5 of the DWARF symbols, which is currently not the default, is another trick to help us start debugging faster. With it, certain information is stored in the main binary that the default version 4 left out. Specifically, we can determine the full set of source files just from the main binary. This allows the debugger to do basic actions like showing the full source tree and setting breakpoints without loading and parsing the full symbol data. This makes debugging with split symbols a lot faster, so we're always using the `-gsplit-dwarf` and `-gdwarf-5` command line flags together!
+
+WIth the DWARF5 debug format we also get access to another useful feature. It introduces a  *name index* in the debug data that will be generated when passing the `-gpubnames` flag:
```
Comment (pfaffe) [inline site/en/blog/faster-wasm-debugging/index.md:190]: There's a capital in wIth.

**Review by pfaffe — APPROVED** [review]

File: site/en/blog/faster-wasm-debugging/index.md [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+
+In the last post, we discussed how to compile and debug this example. Let's do that again, but let's also take a peek at the _//performance//_:
+
+```bash
+$ emcc -sUSE_SDL=2 -g -O0 -o mandelbrot.html mandelbrot.cc -sALLOW_MEMORY_GROWTH
+```
+
+This command produces a 3MB wasm binary. And the bulk of that, as you might expect, is debug information. You can verify this with the `llvm-objdump` tool <sup>[[1]](#footnote-1)</sup> for example:
+
+<sup id="footnote-1">1</sup>: If you don't have a recent version of `llvm-objdump` on your system, and you are using `emsdk`, you can find it in the `emsdk/upstream/bin` directory.
```
Comment (jecfish) [inline site/en/blog/faster-wasm-debugging/index.md:91]:
```suggestion
```

File: site/en/blog/faster-wasm-debugging/index.md [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+
+
+### TL;DR: What’s the advantage of using debug fission?
+
+There are several advantages to splitting up the debug information if one is working with big applications:
+
+1. **Faster linking: The linker no longer needs to parse the entire debug information**. Linkers usually need to parse the entire DWARF data that is in the binary. By stripping out large parts of the debug information into separate files, linkers deal with smaller binaries, which results in faster linking times (especially true for large applications).
+
+2. **Faster debugging: The debugger can skip parsing the additional symbols in `.dwo`/`.dwp` files for some symbol lookups**. For some lookups (such as requests on the line mapping of wasm-to-C++ files), we don’t need to look into the additional debug data. This saves us time, not needing to load the and parse the additional debug data.
+
```
Comment (jecfish) [inline site/en/blog/faster-wasm-debugging/index.md:228]:
```suggestion

<sup id="footnote-1">1</sup>: If you don't have a recent version of `llvm-objdump` on your system, and you are using `emsdk`, you can find it in the `emsdk/upstream/bin` directory.
```

**Pull request merged**